src/NNCache.h

```diff
@@ -0,0 +1,72 @@
+/*
+    This file is part of Leela Zero.
+    Copyright (C) 2017 Gian-Carlo Pascutto
+
+    Leela Zero is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    Leela Zero is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with Leela Zero.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef NNCACHE_H_INCLUDED
+#define NNCACHE_H_INCLUDED
+
+#include "config.h"
+
+#include <vector>
+#include <deque>
+#include <unordered_map>
+
+#include "Network.h"
+#include "SMP.h"
+
+class NNCache {
+public:
+    // return the global NNCache
+    static NNCache* get_NNCache(void);
+
+    // Try and find an existing entry.
+    const Network::Netresult* lookup(const Network::NNPlanes& features);
+
+    // Insert a new entry.
+    void insert(const Network::NNPlanes& features, const Network::Netresult& result);
+
+    // Return the hit rate ratio.
+    std::pair<int, int> hit_rate() const { return {m_hits, m_lookups}; }
+
+    void dump_stats();
+
+private:
+    NNCache(int size = 50000);  // ~ 200MB
+
+    SMP::Mutex m_mutex;
+
+    size_t m_size;
+
+    // Statistics
+    int m_hits{0};
+    int m_lookups{0};
+    int m_inserts{0};
+    int m_collisions{0};
+
+    struct Entry {
+        Entry(const Network::NNPlanes& f, const Network::Netresult& r) : features(f), result(r) {}
+        Network::NNPlanes features; // ~ 1KB
```

Comment:
Some food for thought, From my experience with the Chess engines most engines they don't store the position date in the TTable (this isn't directly compatible because the have really small TTable entries) as it's expense to store more bits). Given you only store 50,000 positions a hash collision on a 64 bit hash is fairly unlikely (I'll do the birthday paradox math later) (and it's not catastrophic if we return bad eval data) it might be faster and more space efficient to not store and compare each time.

Reply:
The question is whether his hash has properties that are as good as typical Zobrist hashes. I don't think that's a given.

Reply:
If i read http://preshing.com/20110504/hash-collision-probabilities/ correctly,  it's 1/10 billion my own math to follow.

We have something slightly different.
Assume NNCache is 200K and full, for a 1000 move game where we do 100K playouts each move we'd have 1e8 lookups (assuming each position is unique AKA this never returns a hit) each with a 2^64/2e5 probability of collision => 2e13/2^64 = one in five million chance of collision (assuming hash is uniformly distributed) during the entire game

Reply:
@Gcp wish I was at home to evaluate and don't want to block the merge but always a thing we can investigate later!

Reply:
>assuming hash is uniformly distributed

Right that's the problem here. Zobrist gives you that, but I'm not positive about this one. I guess the amount of collisions in a long game will tell?

Reply:
The key (~1KB) is relatively small compared to the value (~3KB), so there's only small memory savings in dropping it entirely.

Reply:
In addition to memory saving it makes cache lookup faster (in case of hit) because you don't have to perform the comparison

Reply:
Amused: I think you're reaching here ;) The comparison is much less than 1% of CPU.

FWIW: On my machine with a relatively fast GPU (Titan X), all the CPU operations take less than 10% of wall time. The actual CPU consumption of leelaz is 100%, but more than 90% of that is the terrible OpenCL busy-wait in clFinish().

As far as I can tell, it's basically useless to optimize anything on the CPU at the moment.
I recommend doing the simple reliable thing here and optimizing later if it proves to be worth it.

Reply:
Don't want to derail this pull request; I'm a huge fan and excited to see it going in.

I totally agree with you, the performance isn't big enough to warrant worrying about (at least right now). And it's a good self-consistency check / detector for bad hashing / overclocked computers / non-random bitflips.

Just adding some data because I was curious (and had free time),
https://gist.github.com/sethtroisi/76aeffe8b6b7a29ea4b3eabd49ca46c1

I tested hashing all `bitset<121>` with up to 5 bits set (197 million entries, 9 gigs of ram) for collisions. I didn't find any collisions, which is a good sign that gcc7, at least, uses a uniform hash (as is required by the [spec](http://en.cppreference.com/w/cpp/utility/hash)).

Inserting 197 million entries means 4e16 chances for collisions.
if `hasher<bitset<121>>` is uniform the chance of collision for a 64 bit hash is `4e16/2^64 ~= 2/1000 chance of collision`. if the hash was not uniform (even by a factor or 4) it would be much more likely to have see a collision.

I used the lower 32 bits from the hash and predictable see collisions after ~50-60 thousand insertions.

Additionally 4e16 collisions is greater than the number of comparisons that would be expected from generating 1 million self-play games (1600 node expansions * 65% cache miss * 400 moves/game * 50000 hashes to collide with * 1,000,000 self play games ~= 2e16 chances for collisions)